packages/aggregator/src/analytics/aggregateAnalytics/adjustOptionsToAggregationList.js

```diff
@@ -0,0 +1,134 @@
+  dataSourceEntityType;
+
+const shouldFetchRelations = ({
+  type,
+  config: { aggregationEntityType, dataSourceEntityType } = {},
+} = {}) => aggregationEntityType && dataSourceEntityType && !(type === AGGREGATION_TYPES.RAW);
+
+const getAdjustedOrganisationUnitsAndAggregations = async (
+  context,
+  fetchOptions,
+  aggregationList,
+) => {
+  const { hierarchy, organisationUnitCodes } = fetchOptions;
+  if (!aggregationList.some(shouldFetchDataSourceEntities)) {
+    return [organisationUnitCodes, aggregationList];
+  }
+
+  const { session } = context;
+
+  let adjustedOrganisationUnitCodes;
+  const adjustedAggregations = [];
+  const entityConnection = new EntityConnection(session);
+  await Promise.all(
+    aggregationList.map(async aggregation => {
+      if (!shouldFetchDataSourceEntities(aggregation)) {
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      if (!adjustedOrganisationUnitCodes && !shouldFetchRelations(aggregation)) {
+        const { dataSourceEntityType, dataSourceEntityFilter } = aggregation.config;
+        adjustedOrganisationUnitCodes = await entityConnection.getDataSourceEntities(
+          hierarchy,
+          organisationUnitCodes,
+          dataSourceEntityType,
+          dataSourceEntityFilter,
+        );
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      const {
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      } = aggregation.config;
+      const [
+        dataSourceEntities,
+        relations,
+      ] = await entityConnection.getDataSourceEntitiesAndRelations(
+        hierarchy,
+        organisationUnitCodes,
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      );
+
+      if (!adjustedOrganisationUnitCodes) {
+        adjustedOrganisationUnitCodes = dataSourceEntities;
+      }
+      adjustedAggregations.push({
+        ...aggregation,
+        config: { ...aggregation.config, orgUnitMap: relations },
+      });
+    }),
+  );
+
+  return [adjustedOrganisationUnitCodes, adjustedAggregations];
+};
+
+/**
+ * Each aggregation in the list may require the original fetch option dimensions (dates, org units etc)
+ * to be adjusted, eg when we need to take into account past/future data
+ *
+ * @param {Object[]} aggregationList
+ */
+export const adjustOptionsToAggregationList = async (context, fetchOptions, aggregationOptions) => {
+  const { aggregations: aggregationList = [] } = aggregationOptions;
+  if (aggregationList.length === 0) {
+    return [fetchOptions, aggregationOptions];
+  }
+
+  const { startDate, endDate } = getAdjustedDateRange(fetchOptions, aggregationList);
+  const period = getAdjustedPeriod(fetchOptions, { startDate, endDate });
+
+  const [organisationUnitCodes, aggregations] = await getAdjustedOrganisationUnitsAndAggregations(
```

Comment:
I'd expect to see an object returned here, but you know what - I actually don't mind the array, since `getAdjustedOrganisationUnitsAndAggregations` is local to this file.

I find that although not often used in JavaScript, array return values can make things simpler in some specific cases. So, I'm personally happy with this one 😃

Reply:
I've been using tuples a lot lately. Personally I've been enjoying them, but I can totally see the downsides...

Reply:
They are also convenient as test input, and TypeScript makes them easier to use - since you can define/see the type for each element, which adds context to them.